docs/spec/constructors.rst

```diff
@@ -0,0 +1,487 @@
+    assert_type(MyClass(1.0), MyClass[float])
+
+If any class-scoped type variables are not solved when evaluating the ``__new__``
+method call using the supplied arguments, these type variables should be left
+unsolved, allowing the ``__init__`` method (if applicable) to be used to solve
+them.
+
+  ::
+
+      class MyClass[T]:
+          def __new__(cls, *args, **kwargs) -> Self:
+              return super().__new__(cls)
+
+          def __init__(self, x: T) -> None:
+              pass
+
+      assert_type(MyClass(1), MyClass[int])
+      assert_type(MyClass(""), MyClass[str])
+
+For most classes, the return type for the ``__new__`` method is typically
+``Self``, but other types are also allowed. For example, the ``__new__``
+method may return an instance of a subclass or an instance of some completely
+unrelated class.
+
+If the evaluated return type of ``__new__`` is not an instance of the class
+being constructed (or a subclass thereof) or is a union that includes such
+a class, a type checker should assume that the ``__init__`` method will not be
+called. This is consistent with the runtime behavior of the ``type.__call__``
+method.
+
+  ::
+
+    class MyClass:
+        def __new__(cls) -> int:
+            return 0
+
+        # The __init__ method will not be called in this case, so
+        # it should not be evaluated.
+        def __init__(self, x: int):
+            pass
+
+    assert_type(MyClass(), int)
+
+For purposes of this test, an explicit return type of ``Any`` (or a
+union containing ``Any``) should be treated as a type that is not an instance
+of the class being constructed.
+
+  ::
+
+    class MyClass:
+        def __new__(cls) -> Any:
+            return 0
+
+        # The __init__ method will not be called in this case, so
+        # it should not be evaluated.
+        def __init__(self, x: int):
+            pass
+
+    assert_type(MyClass(), Any)
+
+If the return type of ``__new__`` is not annotated, a type checker may assume
+that the return type is ``Self`` and proceed with the assumption that the
+``__init__`` method will be called.
+
+If the class is generic, it is possible for a ``__new__`` method to override
+the specialized class type and return a class instance that is specialized
+with different type arguments.
+
+  ::
+
+    class MyClass[T]:
+        def __new__(cls, *args, **kwargs) -> "MyClass[list[T]]":
+            ...
+
+    assert_type(MyClass[int](), MyClass[list[int]])
+
+If the ``cls`` parameter within the ``__new__`` method is not annotated, type
+checkers should infer a type of ``type[Self]``. Regardless of whether the
+type of the ``cls`` parameter is explicit or inferred, the type checker should
+bind the class being constructed to this parameter and report any type errors
+that arise during binding.
+
+  ::
+
+    class MyClass[T]:
+        def __new__(cls: "type[MyClass[int]]") -> "MyClass[int]": ...
+
+    MyClass()  # OK
+    MyClass[int]()  # OK
+    MyClass[str]()  # Type Error
```

Comment:
This typically arises when overloads are involved. The example doesn't show this (because it would add undue complexity to the sample), but the fact that it's a type error is important to note because it affects overload matching.